linker/shared/src/main/scala/org/scalajs/linker/backend/emitter/FunctionEmitter.scala

```diff
@@ -2944,12 +2953,27 @@ private[emitter] class FunctionEmitter(sjsGen: SJSGen) {
               tree.getClass)
       }
 
-      if (preserveChar || tree.tpe != CharType)
+      if (preserveChar || realTreeType(tree) != CharType)
         baseResult
       else
         genCallHelper("bC", baseResult)
     }
 
+    private def realTreeType(tree: Tree)(implicit env: Env): Type = {
```

Comment:
This highlights an actual issue with `This().tpe` being lenient. There is already a TODO in the `ClassDefChecker` at
https://github.com/scala-js/scala-js/blob/112def5cb59b942d8b2c4c5791b17ea0c7c4a7dc/linker/shared/src/main/scala/org/scalajs/linker/checker/ClassDefChecker.scala#L757-L758
suggesting that we should enforce a precise type. That makes two reasons now.

I will probably follow-up to this PR afterwards to enforce that, with a deserialization hack for compatibility. That would remove the need for this `realTreeType`.